File: c++/src/xdevs/core/modeling/Coupled.cpp
/*
 * Coupled.cpp
 *
 *  Created on: 5 de may. de 2016
 *      Author: jlrisco
 */

#include "Coupled.h"

Coupled::Coupled(const std::string& name) : Component(name), components(), ic(), eic(), eoc() {
}

Coupled::~Coupled() {
	for(auto coupling : eic) {
		delete coupling;
	}
	for(auto coupling : eoc) {
		delete coupling;
	}
	for(auto coupling : ic) {
		delete coupling;
	}
}

void Coupled::initialize() {}
void Coupled::exit() {}

void Coupled::addCoupling(Component* cFrom, Port* pFrom, Component* cTo, Port* pTo) {
	Coupling* coupling = new Coupling(pFrom, pTo);
	// Add to connections
	if (cFrom == this) {
		eic.push_back(coupling);
	} else if (cTo == this) {
		eoc.push_back(coupling);
	} else {
		ic.push_back(coupling);
	}
}

void Coupled::addComponent(Component* component) {
	components.push_back(component);
}

const std::list<Component*>& Coupled::getComponents() const {
	return components;
}

const std::list<Coupling*>& Coupled::getIC() const { return ic; }
const std::list<Coupling*>& Coupled::getEIC() const { return eic; }
const std::list<Coupling*>& Coupled::getEOC() const { return eoc; }

Component* Coupled::compFromPath(std::string path) {
    //std::cout << "compFromPath: " << path << std::endl;
    Component* currComp = this;
    Component* nextComp;

    size_t last = 0, next = 0;
    std::string term;
    while ((next = path.find(".", last)) != std::string::npos) {
        term = path.substr(last, next-last);
        nextComp = nullptr;

        if(dynamic_cast<Coupled*>(currComp) == nullptr) return nullptr;

        for(Component* comp: ((Coupled*) currComp)->getComponents()) {
            if(comp->getName() == term) {
                nextComp = comp;
                break;
            }
        }

        if(nextComp == nullptr) return nullptr;
        currComp = nextComp;
        last = next;
    }

    if(dynamic_cast<Coupled*>(currComp) == nullptr) return nullptr;

    for(Component* comp: ((Coupled*) currComp)->getComponents()) {
        term = path.substr(last);
        if(comp->getName() == term) {
            //std::cout << "found: " << path << std::endl;
            return comp;
        }
    }
    return nullptr;
}

Port* Coupled::portFromPath(std::string path) {
    //std::cout << "portFromPath: " << path << std::endl;
    size_t pos = path.rfind(".");

    if(pos == std::string::npos)
        return nullptr;

    std::string compPath = path.substr(0, pos);
    std::string portName = path.substr(pos+1);

    Component* comp = compFromPath(compPath);

    if(comp != nullptr) {
        for(Port* port: comp->getInPorts()) {
            if(port->getName() == portName) {
            std::cout << "port found: " << port->getName() << std::endl;
                return port;
            }
        }

        for(Port* port: comp->getOutPorts()) {
            if(port->getName() == portName) {
                std::cout << "port found: " << port->getName() << std::endl;
                return port;
            }
        }
    }

    return nullptr;
}

File: modeling/coupled.go
package modeling

type CoupledInterface interface {
	ComponentInterface
	AddCoupling(pFrom *Port, pTo *Port)
	GetComponent() ComponentInterface
	GetComponents() []ComponentInterface
	GetComponentByName(name string) ComponentInterface
	AddComponent(comp ComponentInterface)
	GetIC() []*Coupling
	GetEIC() []*Coupling
	GetEOC() []*Coupling
}

type Coupled struct {
	Component
	components []ComponentInterface
	ic, eic, eoc []*Coupling  // TODO map?
}

func NewCoupled(name string) *Coupled {
	c := Coupled{*NewComponent(name), nil, nil, nil, nil}
	return &c
}



func (c *Coupled) Initialize() { }

func (c *Coupled) Exit() { }

func (c *Coupled) AddCoupling(pFrom *Port, pTo *Port) {
	for _, p := range []*Port{pFrom, pTo} {
		if p.GetParent() == nil {
			panic("Port " + pFrom.String()  + " does not have a parent component")
		}
 	}
	coup := NewCoupling(pFrom, pTo)
	if pFrom.GetParent() == &c.Component {
		c.eic = append(c.eic, coup)
	} else if pTo.GetParent() == & c.Component {
		c.eoc = append(c.eoc, coup)
	} else {
		c.ic = append(c.ic, coup)
	}
}

func (c *Coupled) GetComponent() ComponentInterface {
	return &c.Component
}

func (c *Coupled) GetComponents() []ComponentInterface {
	return c.components
}

func (c *Coupled) GetComponentByName(name string) ComponentInterface {
	for _, c := range c.components {
		if c.GetName() == name {
			return c
		}
	}
	panic("no child component has name " + name)
}

func (c *Coupled) AddComponent(comp ComponentInterface) {
	comp.setParent(&c.Component)
	c.components = append(c.components, comp)
}

func (c *Coupled) GetIC() []*Coupling {
	return c.ic
}

func (c *Coupled) GetEIC() []*Coupling {
	return c.eic
}

func (c *Coupled) GetEOC() []*Coupling {
	return c.eoc
}
